WIP: SIMSGEO integration

florianwechsung opened the pull request:
I started moving simsgeo over into simsopt. Currently there seems to be an issue with compiling on macs, running 

```
python3 setup.py build; python3 setup.py install
```

and then `python3 -c 'import simsopt'`

results in `Fatal Python error: PyMUTEX_LOCK(_PyRuntime.ceval.gil.mutex) failed`

@mbkumar suggested that this seems to be due a wrong flag passed to a compiler. My guess is that this happens somewhere in skbuild, since the code compiles and tests fine in the [simsgeo repo](https://github.com/hiddenSymmetries/simsgeo) (or maybe I missed something when moving things over, that's also a possibility)

landreman commented:
I've tried building and loading this branch a bunch of ways, but every approach I've tried fails. When I try building on my macbook using g++ 8.4, I get a similar error to Florian, thought the detailed message is a little different:
~~~~
(20210104-01) ~/simsopt$ python
Python 3.8.5 (default, Sep  4 2020, 02:22:02) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import simsopt
Fatal Python error: PyMUTEX_LOCK(gil->mutex) failed
Python runtime state: unknown

Abort trap: 6
~~~~
I also tried building on my macbook using AppleClang 10.0.0, but the build step fails, as follows:
~~~~
(20210104-01) ~/simsopt$ python setup.py build


--------------------------------------------------------------------------------
-- Trying "Ninja" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- The C compiler identification is AppleClang 10.0.0.10001145
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/mattland/simsopt/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Ninja" generator - success
Configuring Project
  Working directory:
    /Users/mattland/simsopt/_skbuild/macosx-10.14-x86_64-3.8/cmake-build
  Command:
    cmake /Users/mattland/simsopt -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/Users/mattland/simsopt/_skbuild/macosx-10.14-x86_64-3.8/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/Users/mattland/opt/miniconda3/envs/20210104-01/bin/python -DPYTHON_VERSION_STRING:STRING=3.8.5 -DPYTHON_INCLUDE_DIR:PATH=/Users/mattland/opt/miniconda3/envs/20210104-01/include/python3.8 -DPYTHON_LIBRARY:FILEPATH=/Users/mattland/opt/miniconda3/envs/20210104-01/lib/libpython3.8.dylib -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/Users/mattland/opt/miniconda3/envs/20210104-01/lib/python3.8/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release -DCMAKE_OSX_DEPLOYMENT_TARGET:STRING=10.14 -DCMAKE_OSX_ARCHITECTURES:STRING=x86_64

-- The C compiler identification is AppleClang 10.0.0.10001145
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /Users/mattland/opt/miniconda3/envs/20210104-01/bin/python (found suitable version "3.8.5", minimum required is "3") 
-- Found PythonLibs: /Users/mattland/opt/miniconda3/envs/20210104-01/lib/libpython3.8.dylib (found suitable version "3.8.5", minimum required is "3") 
-- Found PythonInterp: /Users/mattland/opt/miniconda3/envs/20210104-01/bin/python (found version "3.8.5") 
-- Found NumPy: version "1.19.1" /Users/mattland/opt/miniconda3/envs/20210104-01/lib/python3.8/site-packages/numpy/core/include
-- Submodule update
Submodule 'thirdparty/blaze' (https://bitbucket.org/blaze-lib/blaze.git) registered for path 'thirdparty/blaze'
Submodule 'thirdparty/pybind11' (https://github.com/pybind/pybind11.git) registered for path 'thirdparty/pybind11'
Submodule 'thirdparty/xsimd' (https://github.com/xtensor-stack/xsimd.git) registered for path 'thirdparty/xsimd'
Submodule 'thirdparty/xtensor' (https://github.com/xtensor-stack/xtensor.git) registered for path 'thirdparty/xtensor'
Submodule 'thirdparty/xtensor-python' (https://github.com/xtensor-stack/xtensor-python.git) registered for path 'thirdparty/xtensor-python'
Submodule 'thirdparty/xtl' (https://github.com/xtensor-stack/xtl.git) registered for path 'thirdparty/xtl'
Cloning into '/Users/mattland/simsopt/thirdparty/blaze'...
Cloning into '/Users/mattland/simsopt/thirdparty/pybind11'...
Cloning into '/Users/mattland/simsopt/thirdparty/xsimd'...
Cloning into '/Users/mattland/simsopt/thirdparty/xtensor'...
Cloning into '/Users/mattland/simsopt/thirdparty/xtensor-python'...
Cloning into '/Users/mattland/simsopt/thirdparty/xtl'...
Submodule path 'thirdparty/blaze': checked out 'daf5b7cc2ef68d2a0b422594cde63c6c396b5a32'
Submodule path 'thirdparty/pybind11': checked out '028812ae7eee307dca5f8f69d467af7b92cc41c8'
Submodule path 'thirdparty/xsimd': checked out '7b2735f4acfd60604f3371bb82a60d9ab7541b41'
Submodule path 'thirdparty/xtensor': checked out 'e2eefa861f4aeae60c35603f654361e4395d0c7a'
Submodule path 'thirdparty/xtensor-python': checked out 'd6f87cf3739d21c60ea9b193f4ba533c0d956007'
Submodule path 'thirdparty/xtl': checked out 'f850992614250521bc7809e7fd7a87e9e08a12d8'
-- pybind11 v2.6.2 dev1
-- Found PythonLibs: /Users/mattland/opt/miniconda3/envs/20210104-01/lib/libpython3.8.dylib
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Performing Test HAS_FLTO_THIN
-- Performing Test HAS_FLTO_THIN - Success
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    SKBUILD


-- Build files have been written to: /Users/mattland/simsopt/_skbuild/macosx-10.14-x86_64-3.8/cmake-build
[1/5] Building CXX object CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o
FAILED: CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -Dsimsgeopp_EXPORTS -I../../../thirdparty/xtensor/include -I../../../thirdparty/xtensor-python/include -I../../../thirdparty/xsimd/include -I../../../thirdparty/xtl/include -I../../../thirdparty/blaze -I/Users/mattland/opt/miniconda3/envs/20210104-01/lib/python3.8/site-packages/numpy/core/include -I../../../src/simsgeopp -isystem /Users/mattland/opt/miniconda3/envs/20210104-01/include/python3.8 -isystem ../../../thirdparty/pybind11/include -O3 -march=native -O3 -DNDEBUG -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -mmacosx-version-min=10.14 -fPIC -fvisibility=hidden -flto -std=gnu++17 -MD -MT CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o -MF CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o.d -o CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o -c ../../../src/simsgeopp/biot_savart.cpp
../../../src/simsgeopp/biot_savart.cpp:187:66: error: no viable constructor or deduction guide for deduction of template arguments of 'optional'
        auto _dB_by_dX = (dB_by_dX.size() <= i) ? std::nullopt : std::optional(std::move(dB_by_dX[i]));
                                                                 ^
~~~~
followed by many other errors/warnings. I then tried on the HPC system I usually use, Draco at IPP, using the default Intel compilers. Again the build step fails, but with a different error:
~~~~
(base) mlan@draco04:~/simsopt> env CC=icc CXX=icpc python setup.py build


--------------------------------------------------------------------------------
-- Trying "Ninja" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- The C compiler identification is Intel 18.0.5.20180823
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /mpcdf/soft/SLE_12/packages/x86_64/intel/18.0.5/bin/icc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is Intel 18.0.5.20180823
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /mpcdf/soft/SLE_12/packages/x86_64/intel/18.0.5/bin/icpc - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /draco/u/mlan/simsopt/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Ninja" generator - success
--------------------------------------------------------------------------------
Configuring Project
  Working directory:
    /draco/u/mlan/simsopt/_skbuild/linux-x86_64-3.7/cmake-build
  Command:
    cmake /draco/u/mlan/simsopt -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/draco/u/mlan/simsopt/_skbuild/linux-x86_64-3.7/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/bin/python -DPYTHON_VERSION_STRING:STRING=3.7.6 -DPYTHON_INCLUDE_DIR:PATH=/mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/include/python3.7m -DPYTHON_LIBRARY:FILEPATH=/mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/lib/libpython3.7m.so -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/u/mlan/.local/lib/python3.7/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release

-- The C compiler identification is Intel 18.0.5.20180823
-- The CXX compiler identification is Intel 18.0.5.20180823
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /mpcdf/soft/SLE_12/packages/x86_64/intel/18.0.5/bin/icc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /mpcdf/soft/SLE_12/packages/x86_64/intel/18.0.5/bin/icpc - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/bin/python (found suitable version "3.7.6", minimum required is "3") 
-- Found PythonLibs: /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/lib/libpython3.7m.so (found suitable version "3.7.6", minimum required is "3") 
-- Found PythonInterp: /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/bin/python (found version "3.7.6") 
-- Found NumPy: version "1.18.1" /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/lib/python3.7/site-packages/numpy/core/include
-- Submodule update
-- pybind11 v2.6.2 dev1
-- Found PythonLibs: /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/lib/libpython3.7m.so
-- Performing Test HAS_INTEL_IPO
-- Performing Test HAS_INTEL_IPO - Success
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    SKBUILD
-- Build files have been written to: /draco/u/mlan/simsopt/_skbuild/linux-x86_64-3.7/cmake-build
[1/5] Building CXX object CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o
FAILED: CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o 
/mpcdf/soft/SLE_12/packages/x86_64/intel/18.0.5/bin/icpc -Dsimsgeopp_EXPORTS -I../../../thirdparty/xtensor/include -I../../../thirdparty/xtensor-python/include -I../../../thirdparty/xsimd/include -I../../../thirdparty/xtl/include -I../../../thirdparty/blaze -I/mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/lib/python3.7/site-packages/numpy/core/include -I../../../src/simsgeopp -isystem /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/include/python3.7m -isystem ../../../thirdparty/pybind11/include -O3 -march=native -O3 -DNDEBUG -fPIC -fvisibility=hidden -ipo -std=gnu++17 -MD -MT CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o -MF CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o.d -o CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o -c ../../../src/simsgeopp/biot_savart.cpp
In file included from ../../../thirdparty/xtensor/include/xtensor/xexpression.hpp(17),
                 from ../../../thirdparty/xtensor/include/xtensor/xio.hpp(21),
                 from ../../../src/simsgeopp/biot_savart.h(5),
                 from ../../../src/simsgeopp/biot_savart.cpp(1):
../../../thirdparty/xtl/include/xtl/xclosure.hpp(26): error: namespace "std" has no member "add_const_t"
      constexpr std::add_const_t<T>& as_const(T& t) noexcept
                     ^

In file included from ../../../thirdparty/xtensor/include/xtensor/xexpression.hpp(17),
                 from ../../../thirdparty/xtensor/include/xtensor/xio.hpp(21),
                 from ../../../src/simsgeopp/biot_savart.h(5),
                 from ../../../src/simsgeopp/biot_savart.cpp(1):
../../../thirdparty/xtl/include/xtl/xclosure.hpp(26): error: "constexpr" is not valid here
      constexpr std::add_const_t<T>& as_const(T& t) noexcept
      ^

In file included from ../../../thirdparty/xtensor/include/xtensor/xexpression.hpp(17),
                 from ../../../thirdparty/xtensor/include/xtensor/xio.hpp(21),
                 from ../../../src/simsgeopp/biot_savart.h(5),
                 from ../../../src/simsgeopp/biot_savart.cpp(1):
../../../thirdparty/xtl/include/xtl/xclosure.hpp(26): error: expected a ";"
      constexpr std::add_const_t<T>& as_const(T& t) noexcept
                                ^
~~~~
followed by many other errors/warnings. I then tried again on Draco using g++ 9.3.0. This time the build and install steps worked. Trying to import the module in python, I got this error:
~~~~
(base) mlan@draco04:~/simsopt> srun python -c "import simsopt"
srun: job 18373388 queued and waiting for resources
srun: job 18373388 has been allocated resources
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/u/mlan/.local/lib/python3.7/site-packages/simsopt-0.0.1-py3.7-linux-x86_64.egg/simsopt/__init__.py", line 3, in <module>
    from .geo import *
  File "/u/mlan/.local/lib/python3.7/site-packages/simsopt-0.0.1-py3.7-linux-x86_64.egg/simsopt/geo/__init__.py", line 3, in <module>
    from .magneticaxis import *
  File "/u/mlan/.local/lib/python3.7/site-packages/simsopt-0.0.1-py3.7-linux-x86_64.egg/simsopt/geo/magneticaxis.py", line 1, in <module>
    from .curve import JaxCurve, Curve
  File "/u/mlan/.local/lib/python3.7/site-packages/simsopt-0.0.1-py3.7-linux-x86_64.egg/simsopt/geo/curve.py", line 2, in <module>
    import simsgeopp as sgpp
  File "/u/mlan/.local/lib/python3.7/site-packages/simsopt-0.0.1-py3.7-linux-x86_64.egg/simsgeopp/__init__.py", line 1, in <module>
    from .simsgeopp import *
ImportError: libpython3.7m.so.1.0: cannot open shared object file: No such file or directory
~~~~
(srun was needed to avoid an MPI error, since MPI only works on the compute nodes.)

For what compiler and machine have you guys gotten this to work?

florianwechsung commented:
I just looked into this again, and it turns out that my installation was picking up different python version for interpreter, header and library.

# AppleClang 12 on MacOS 10.15.7 
If I do

```

pip3 uninstall simsopt
rm -rf _skbuild
python3 setup.py build
python3 setup.py install
```

the relevant cmake output shows that AppleClang 12 is used and python3.8 is found

```
Configuring Project
  Working directory:
    /Users/fw18/Dropbox/Documents/Uni/PostDoc/code/simsopt/_skbuild/macosx-10.15-x86_64-3.8/cmake-build
  Command:
    cmake /Users/fw18/Dropbox/Documents/Uni/PostDoc/code/simsopt -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/Users/fw18/Dropbox/Documents/Uni/PostDoc/code/simsopt/_skbuild/macosx-10.15-x86_64-3.8/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/usr/local/opt/python@3.8/bin/python3.8 -DPYTHON_VERSION_STRING:STRING=3.8.6 -DPYTHON_INCLUDE_DIR:PATH=/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/include/python3.8 -DPYTHON_LIBRARY:FILEPATH=/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/usr/local/lib/python3.8/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release -DCMAKE_OSX_DEPLOYMENT_TARGET:STRING=10.15 -DCMAKE_OSX_ARCHITECTURES:STRING=x86_64

-- The C compiler identification is AppleClang 12.0.0.12000032
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features-- Detecting C compile features - done-- Detecting CXX compiler ABI info-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /usr/local/opt/python@3.8/bin/python3.8 (found suitable version "3.8.6", minimum required is "3") 
-- Found PythonLibs: /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib (found suitable version "3.8.6", minimum required is "3") 
-- Found PythonInterp: /usr/local/opt/python@3.8/bin/python3.8 (found version "3.8.6") 
-- Found NumPy: version "1.19.0" /usr/local/lib/python3.8/site-packages/numpy/core/include
-- Submodule update
-- pybind11 v2.6.2 dev1
-- Found PythonLibs: /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Performing Test HAS_FLTO_THIN
-- Performing Test HAS_FLTO_THIN - Success
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    SKBUILD
```

I can then `python3 -c 'from simsopt.geo import FourierCurve; print(FourierCurve([0.1], 5).gamma())'` it works fine.

# GCC-10 on MacOS 10.15.7 

If I do

```

pip3 uninstall simsopt
rm -rf _skbuild
env CC=gcc-10 CXX=g++-10 python3 setup.py build 
env CC=gcc-10 CXX=g++-10 python3 setup.py install
```

the output is

```
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Checking whether C compiler has -isysroot
-- Checking whether C compiler has -isysroot - yes
-- Checking whether C compiler supports OSX deployment target flag
-- Checking whether C compiler supports OSX deployment target flag - yes
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/local/bin/gcc-10 - skipped 
-- Detecting C compile features
-- Detecting C compile features - done
-- Checking whether CXX compiler has -isysroot
-- Checking whether CXX compiler has -isysroot - yes
-- Checking whether CXX compiler supports OSX deployment target flag
-- Checking whether CXX compiler supports OSX deployment target flag - yes
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/bin/g++-10 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /usr/local/opt/python@3.8/bin/python3.8 (found suitable version "3.8.6", minimum required is "3") 
-- Found PythonLibs: /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib (found suitable version "3.8.6", minimum required is "3") 
-- Found PythonInterp: /usr/local/opt/python@3.8/bin/python3.8 (found version "3.8.6") 
-- Found NumPy: version "1.19.0" /usr/local/lib/python3.8/site-packages/numpy/core/include
-- Submodule update
-- pybind11 v2.6.2 dev1
-- Found PythonLibs: /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/libpython3.8.dylib
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Failed
-- Configuring done
-- Generating done
```

and I can again `python3 -c 'from simsopt.geo import FourierCurve; print(FourierCurve([0.1], 5).gamma())'` without problems.

C++ `optional` is a fairly recent (C++17) feature, and was only added to xcode with version 11 (see [here](https://developer.apple.com/documentation/xcode-release-notes/xcode-11-release-notes)) So that should explain your issue on Mac with Clang 10. Let's maybe try to figure out why it doesn't work with gcc8 on your mac, before we try to understand the cluster. can you post the output during the build phase when you use gcc? Just to make sure everything is clean, please run `pip3 uninstall simsopt ` as many times as it takes until pip says it can't find simsopt anymore (sometimes multiple versions of a package are installed), and also delete `_skbuild`.

landreman commented:
Thanks @florianwechsung for the info. I tried some updated compiler versions on my macbook:

clang
-----
My OS is 10.14.6, and I'd prefer not to update to >= 10.15 since there is a piece of software I use (Mathtype) that does not yet work for 10.15. The most recent version of xcode compatible with OS 10.14 [appears to be](https://xcodereleases.com/) xcode 11.3.1, so I installed that, which provides clang 11.0.0. Now `python setup.py build` and `python setup.py install` succeed, without that `std::optional` error from clang 10. However, when I try to import in python I get that "PyMUTEX" error:
~~~~
(20210104-01) ~/simsopt$ python
Python 3.8.5 (default, Sep  4 2020, 02:22:02) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import simsopt
Fatal Python error: PyMUTEX_LOCK(gil->mutex) failed
Python runtime state: unknown

Abort trap: 6
~~~~

GCC
----
I updated to gcc 10.2.0, the same version as you, but get the same "PyMUTEX" error as before. I build using
~~~~
env CC=gcc-mp-10 CXX=g++-mp-10 python setup.py build
~~~~
and the cmake output shows
~~~~
Configuring Project
  Working directory:
    /Users/mattland/simsopt/_skbuild/macosx-10.14-x86_64-3.8/cmake-build
  Command:
    cmake /Users/mattland/simsopt -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/Users/mattland/simsopt/_skbuild/macosx-10.14-x86_64-3.8/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/Users/mattland/opt/miniconda3/envs/20210104-01/bin/python -DPYTHON_VERSION_STRING:STRING=3.8.5 -DPYTHON_INCLUDE_DIR:PATH=/Users/mattland/opt/miniconda3/envs/20210104-01/include/python3.8 -DPYTHON_LIBRARY:FILEPATH=/Users/mattland/opt/miniconda3/envs/20210104-01/lib/libpython3.8.dylib -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/Users/mattland/opt/miniconda3/envs/20210104-01/lib/python3.8/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release -DCMAKE_OSX_DEPLOYMENT_TARGET:STRING=10.14 -DCMAKE_OSX_ARCHITECTURES:STRING=x86_64

-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Checking whether C compiler has -isysroot
-- Checking whether C compiler has -isysroot - yes
-- Checking whether C compiler supports OSX deployment target flag
-- Checking whether C compiler supports OSX deployment target flag - yes
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/local/bin/gcc-mp-10 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Checking whether CXX compiler has -isysroot
-- Checking whether CXX compiler has -isysroot - yes
-- Checking whether CXX compiler supports OSX deployment target flag
-- Checking whether CXX compiler supports OSX deployment target flag - yes
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/local/bin/g++-mp-10 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /Users/mattland/opt/miniconda3/envs/20210104-01/bin/python (found suitable version "3.8.5", minimum required is "3") 
-- Found PythonLibs: /Users/mattland/opt/miniconda3/envs/20210104-01/lib/libpython3.8.dylib (found suitable version "3.8.5", minimum required is "3") 
-- Found PythonInterp: /Users/mattland/opt/miniconda3/envs/20210104-01/bin/python (found version "3.8.5") 
-- Found NumPy: version "1.19.1" /Users/mattland/opt/miniconda3/envs/20210104-01/lib/python3.8/site-packages/numpy/core/include
-- Submodule update
-- pybind11 v2.6.2 dev1
-- Found PythonLibs: /Users/mattland/opt/miniconda3/envs/20210104-01/lib/libpython3.8.dylib
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Failed
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    SKBUILD
~~~~

Any chance this PyMUTEX error is fix-able without me updating to OS 10.15?

Intel compiler on IPP-Draco
---------------------------
Just a comment about the error from my earlier message. The error says `namespace "std" has no member "add_const_t"`. But `std::add_const_t<T>` [appears to be in the c++14 standard onward](https://en.cppreference.com/w/cpp/types/add_cv), and the compile command did include `-std=g++17`. So I'm confused.

florianwechsung commented:
Mh, let's try to isolate the problem.

After doing `pip3 uninstall simsopt simsgeo simsgeopp` a couple times (until none of these are installed), can you try
```
mkdir build
cd build
cmake ..
make -j
ipython3
> import simsgeopp
```

landreman commented:
@florianwechsung I tried these steps you suggested for building using cmake directly. For both clang11 and gcc10.2.0, at the import step at the end, I get the same `PyMUTEX_LOCK` error as before, when using the python installation I've been using, 3.8.5 from miniconda. I tried again after `conda deactivate` so as to instead use python 3.7.9 from macports. With that change, the simsopt build and import now works with both clang and gcc. Hooray! I don't think the `PyMUTEX_LOCK` is related to the python version, since you found simsopt worked with python 3.8.6 on your machine. Maybe the issue is related to conda, or some quirk of the packages on my laptop. It would be nice if I could get simsopt working with conda so I can try different combinations of packages and package versions.

florianwechsung commented:
Can you uncomment these lines in the CMakeLists.txt
```
get_cmake_property(_variableNames VARIABLES)
list (SORT _variableNames)
foreach (_variableName ${_variableNames})
    message(STATUS "${_variableName}=${${_variableName}}")
endforeach()

```

(that will print out all variables that CMake knows of) and then run 

`cmake .. | grep python` in a clean `build` directory. I wonder if it uses a wrong `Python.h` from somewhere else.

landreman commented:
Here is the result, using my conda python 3.8.5 installation with clang11. It looks like all the python-related variables correctly point to the same place.
[cmake_output.txt](https://github.com/hiddenSymmetries/simsopt/files/5783997/cmake_output.txt)
[cmake_output_grep_python.txt](https://github.com/hiddenSymmetries/simsopt/files/5784008/cmake_output_grep_python.txt)

At the start of this pull request, you were getting a similar (but not identical) PyMUTEX_LOCK error, whereas in your Jan 4th message you were not. What exactly changed on your end in between?

florianwechsung commented:
I was using pyenv before to select a different python version, and noticed that PythonInterp and PythonLibraries were not pointing to the same installation. Turning off pyenv fixed that. The few results I find when googling this error also seem to be related to different python installations  being picked up, hence my suggestion.

Do you remember by any chance whether you were using conda back when we compiled PyPlasmaOpt on your machine?

landreman commented:
When we compiled PyPlasmaOpt on my macbook, it was before I started using conda.

florianwechsung commented:
@landreman I was able to reproduce the issue on my mac with miniconda. Commenting out 

```
target_link_libraries(simsgeopp PUBLIC ${PYTHON_LIBRARY})
```
fixed the problem for me. Can you try that?

florianwechsung commented:
I don't fully understand why this fixes it, but looking at the definition of `pybind11_add_module`, some linking already happens inside there, so maybe there is clash.

landreman commented:
@florianwechsung Commenting out that target_link_libraries line worked on my end! Thanks!

landreman commented:
Here's my understanding of where we stand:

- `python setup.py build; python setup.py install` works on GCC and clang, both on my laptop and on the IPP HPC system Draco.
- `pip install .` does not work anywhere: the simsgeopp extension is built fine but it gets installed in the wrong place, several directories up from `site-packages`.
- I've never managed to build with the Intel compiler, even using `python setup.py build`. 

For the `pip install` issue, one solution seems to be eliminating scikit-build, which I did in the [simsgeo-no-scikit-build branch](https://github.com/hiddenSymmetries/simsopt/tree/simsgeo-no-scikit-build). In this branch I added the [boilerplate code recommended by pybind11](https://github.com/pybind/cmake_example/blob/master/setup.py) to `setup.py`. Now `pip install .` works on my macbook and the ci. The tox test isn't passing yet due to not finding `CMakeLists.txt`, but I think this could be fixed. scikit-build could still be used for the VMEC2000 module. Any objections if we go with this scikit-build-free approach for simsopt itself?

Regarding the Intel compiler, I think we should get this working since many HPC systems use the Intel compiler by default (Cori, Draco, PPPL). It appears that `xtl` requires a specific version of the gcc C++ standard library even when compiling with Intel!?! See e.g. [here](https://software.intel.com/content/www/us/en/develop/articles/icpc-error-namespace-std-has-no-member-function-or-bind-or-other.html) and [here](https://www.reddit.com/r/cpp_questions/comments/cex2es/stdget_time_with_intel_c_compiler/). I experience the same behavior even for the [official xtensor-python cookiecutter example](https://github.com/xtensor-stack/xtensor-python-cookiecutter). For `xtl` only gcc8 seems to work, not 9 or 10. This would be a pain for users. Even with gcc8, so xtl works, there are still fatal errors from blaze. For contrast, there are no problems at all with Intel for the [booz_xform package we discussed last week](https://github.com/hiddenSymmetries/booz_xform), which uses pybind11 and Eigen, and doesn't seem to require any gcc library. We may want to consider refactoring to use Eigen instead of blaze/xtensor in order to make an Intel build possible. Any other ideas?

Here is the `xtl` error I get with Intel if gcc is too old:
~~~~
  icpc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/u/mlan/.local/lib/python3.7/site-packages/pybind11/include -I/u/mlan/.local/lib/python3.7/site-packages/pybind11/include -I/u/mlan/.local/lib/python3.7/site-packages/numpy-1.20.0rc2-py3.7-linux-x86_64.egg/numpy/core/include -I/tmp/pip-req-build-14b2df_6/externalPackages/xtl/include -I/tmp/pip-req-build-14b2df_6/externalPackages/xtensor/include -I/tmp/pip-req-build-14b2df_6/externalPackages/xtensor-python/include -I/mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/include -I/mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/Library/include -I/mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/include/python3.7m -c src/main.cpp -o build/temp.linux-x86_64-3.7/src/main.o -DVERSION_INFO="0.0.1" -std=c++14 -fvisibility=hidden
  In file included from /tmp/pip-req-build-14b2df_6/externalPackages/xtl/include/xtl/xcomplex.hpp(30),
                   from /tmp/pip-req-build-14b2df_6/externalPackages/xtensor/include/xtensor/xmath.hpp(23),
                   from src/main.cpp(3):
  /tmp/pip-req-build-14b2df_6/externalPackages/xtl/include/xtl/xclosure.hpp(26): error: namespace "std" has no member "add_const_t"
        constexpr std::add_const_t<T>& as_const(T& t) noexcept
                       ^
~~~~

Here is the error I get with Intel + gcc9:
~~~~
/mpcdf/soft/SLE_12/packages/x86_64/intel/18.0.5/bin/icpc -Dsimsgeopp_EXPORTS -I../../../thirdparty/xtensor/include -I../../../thirdparty/xtensor-python/include -I../../../thirdparty/xsimd/include -I../../../thirdparty/xtl/include -I../../../thirdparty/blaze -I/u/mlan/.local/lib/python3.7/site-packages/numpy-1.20.0rc2-py3.7-linux-x86_64.egg/numpy/core/include -I../../../src/simsgeopp -isystem /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/include/python3.7m -isystem ../../../thirdparty/pybind11/include -O3 -march=native -O3 -DNDEBUG -fPIC -fvisibility=hidden -ipo -qopenmp -std=gnu++17 -MD -MT CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o -MF CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o.d -o CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart.cpp.o -c ../../../src/simsgeopp/biot_savart.cpp
In file included from /draco/u/system/soft/SLE_12/packages/x86_64/gcc/9.3.0/bin/../include/c++/9.3.0/bits/ios_base.h(46),
                 from /draco/u/system/soft/SLE_12/packages/x86_64/gcc/9.3.0/bin/../include/c++/9.3.0/ios(42),
                 from /draco/u/system/soft/SLE_12/packages/x86_64/gcc/9.3.0/bin/../include/c++/9.3.0/istream(38),
                 from /draco/u/system/soft/SLE_12/packages/x86_64/gcc/9.3.0/bin/../include/c++/9.3.0/sstream(38),
                 from /draco/u/system/soft/SLE_12/packages/x86_64/gcc/9.3.0/bin/../include/c++/9.3.0/complex(45),
                 from /draco/u/system/soft/SLE_12/packages/x86_64/intel_parallel_studio/2018.4/compilers_and_libraries_2018.5.274/linux/compiler/include/complex(30),
                 from ../../../thirdparty/xtensor/include/xtensor/xio.hpp(13),
                 from ../../../src/simsgeopp/biot_savart.h(5),
                 from ../../../src/simsgeopp/biot_savart.cpp(1):
/draco/u/system/soft/SLE_12/packages/x86_64/gcc/9.3.0/bin/../include/c++/9.3.0/bits/stl_function.h(437): error: identifier "__builtin_is_constant_evaluated" is undefined
  	if (__builtin_is_constant_evaluated())
  	    ^
~~~~
Here is the blaze error I get with Intel + gcc8:
~~~~/mpcdf/soft/SLE_12/packages/x86_64/intel/19.1.3/bin/icpc -Dsimsgeopp_EXPORTS -I../../../thirdparty/xtensor/include -I../../../thirdparty/xtensor-python/include -I../../../thirdparty/xsimd/include -I../../../thirdparty/xtl/include -I../../../thirdparty/blaze -I/u/mlan/.local/lib/python3.7/site-packages/numpy-1.20.0rc2-py3.7-linux-x86_64.egg/numpy/core/include -I../../../src/simsgeopp -isystem /mpcdf/soft/SLE_12/packages/x86_64/anaconda/3/2020.02/include/python3.7m -isystem ../../../thirdparty/pybind11/include -O3 -march=native -O3 -DNDEBUG -fPIC -fvisibility=hidden -ipo -qopenmp -std=gnu++17 -MD -MT CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart_derivative.cpp.o -MF CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart_derivative.cpp.o.d -o CMakeFiles/simsgeopp.dir/src/simsgeopp/biot_savart_derivative.cpp.o -c ../../../src/simsgeopp/biot_savart_derivative.cpp
In file included from ../../../thirdparty/blaze/blaze/math/smp/default/DenseVector.h(43),
                 from ../../../thirdparty/blaze/blaze/math/smp/DenseVector.h(52),
                 from ../../../thirdparty/blaze/blaze/math/Band.h(48),
                 from ../../../thirdparty/blaze/blaze/Math.h(46),
                 from ../../../thirdparty/blaze/blaze/Blaze.h(43),
                 from ../../../src/simsgeopp/biot_savart.h(8),
                 from ../../../src/simsgeopp/biot_savart_derivative.cpp(1):
../../../thirdparty/blaze/blaze/math/expressions/Vector.h(269): warning #3058: GNU attributes on a template redeclaration have no effect
  BLAZE_ALWAYS_INLINE VT& crtp_cast( Vector<VT,TF>& vector )
  ^
…
In file included from ../../../thirdparty/blaze/blaze/math/CompressedVector.h(43),
                 from ../../../thirdparty/blaze/blaze/math/CompressedMatrix.h(46),
                 from ../../../thirdparty/blaze/blaze/Math.h(48),
                 from ../../../thirdparty/blaze/blaze/Blaze.h(43),
                 from ../../../src/simsgeopp/biot_savart.h(8),
                 from ../../../src/simsgeopp/biot_savart_derivative.cpp(1):
../../../thirdparty/blaze/blaze/math/dense/StaticVector.h(416): error: expected a ")"
          IsSIMDCombinable_v< Type, ElementType_t<VT> > );
                                                         ^
~~~~

florianwechsung commented:
> Any objections if we go with this scikit-build-free approach for simsopt itself?

No objections from my side.

> We may want to consider refactoring to use Eigen instead of blaze/xtensor in order to make an Intel build possible. Any other ideas?

This is more tricky. I have no objections to swapping blaze for eigen, but getting rid of xtensor won't be so easy. I use xtensor because it allows tensor of arbitrary dimensions. Let's take a vector field evaluated at `n` points for example. Then
- `B` is an array of dimension `n x 3`
- `\nabla B` is an array of dimension `n x 3 x 3`
- `\nabla\nabla B` is an array of dimension `n x 3 x 3 x 3`

AFAIK neither eigen nor blaze have proper support for this, in particular in combination with pybind11.

Before we abandon xtensor, maybe it would be worth posting an issue on the xtensor github page, and see if they have any advice?

landreman commented:
I looked into what is involved in using arbitrary-rank Eigen tensors with pybind11. Here is a working example, passing tensors both from python to C++ and C++ to python:

https://github.com/landreman/eigen_tensors 

It built on the first attempt on every machine and compiler I tried: with gcc (even back to v4.8), clang, and intel, including on Mac, PPPL, Cori, and IPP-Draco. When compiling with Intel on the HPC systems, no gcc module needs to be loaded. To get a numpy array into Eigen, a few lines of code are required:
   https://github.com/landreman/eigen_tensors/blob/3f629022a0a2cb313964f333237ea25906ceff1a/src/eigen_tensors.cpp#L55-L62

Getting the Eigen tensor back to python requires 1 manual line:

https://github.com/landreman/eigen_tensors/blob/3f629022a0a2cb313964f333237ea25906ceff1a/src/eigen_tensors.cpp#L77

If we do those conversions more than once, perhaps they could be put into a function or macro.

So to summarize, it looks like the C++-python conversions are not as slick with Eigen arbitrary-rank tensors as with xtensor, but they are not too bad, and code appears significantly more portable with Eigen.

florianwechsung commented:
Thank you for looking into this!

I think I'm on board with this. Replacing blaze+xtensor+xtensor-python+xtl with just eigen is a clear improvement. I feel slightly uncomfortable about depending on something that is officially unsupported by eigen, but let's hope that doesn't come back to bite us.

The functionality of xsimd is lower level than what eigen does, so I don't think we can get rid of that as easily, but it hasn't caused any problems so far, so hopefully it will stay that way. Worst case we have to switch to one of the competitors of xsimd, such as Vc or libsimdpp. 

I can make the required changes after my simons talk next week.

landreman commented:
I'll throw out another option too. If I'm not mistaken, there are no linear algebra solves / matrix decompositions / eigenvalue operations used in simsgeo, right? If all that is needed is tensor indexing, we don't fundamentally need a linear algebra library like blaze or eigen at all - we could just use homemade classes for tensors that implement indexing. I did this recently for an unrelated project here: https://github.com/landreman/qsc/blob/master/src/vector_matrix.hpp#L106-L154 and https://github.com/landreman/qsc/blob/master/src/vector_matrix.cpp#L187-L260 . I have not tried this approach yet with pybind11. I suspect a little more code would be needed to accept and return numpy arrays than with the eigen approach - not sure. Also I'm not sure if eigen/xtensor use any tricks related to memory alignment or cache sizes to improve the performance of tensor indexing. But the potential advantage of this homemade-tensor approach is elimination of dependencies, and the simplest and fastest build. Downsides are it requires maintaining our own tensor implementation, and we might need linear algebra from eigen/blaze later for some other part of simsopt anyway. So I'm not necessarily advocating this approach, just wanted to put it on the table.

florianwechsung commented:
I think we'll eventually do some linear algebra anyway, so I don't see us relying on eigen or blaze as a problem that makes it worth handrolling our own tensor/matrix/vector code.

One issue that I just noticed is the following: 
In xtensor the rank of the tensor is not part of its type, where as in Eigen the rank is part of the type.
This means that for the cache we can't just do 
`map<string, Tensor<double>>` 
since different tensors have different ranks and are thus of different type.
My fix would be to use `std::variant` (a C++17 feature), which allows us to create a dict that could contain vector, matrices, or tensors, e.g. via

`map<string, variant<Tensor<double, 1>, Tensor<double, 2>, Tensor<double, 3>>> cache`

But I'm not sure this is the most elegant solution. Any other suggestions?

florianwechsung commented:
I replaced blaze with eigen and made a few small changes so that the code now compiles on C++14. I updated both this branch and the `simsgeo-no-scikit-build` branch.

@landreman could you see if this helps with the compilation woes that we were having?

landreman commented:
Thanks for working on this. With these latest commits, it is long last able to compile on IPP-Draco. The default intel/18 compiler there does not work (`"pybind11 supports only C++11 with Intel C++ compiler v18. Use v19 or newer for C++14."`), but intel/19 is available. Since the default gcc is 4.8.5 I also had to load a gcc module (even though using the intel compiler); gcc-8, 9, and 10 all work with intel/19. I'll test on NERSC-Cori too.
